test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs

```diff
@@ -0,0 +1,412 @@
+            var packageVersion2 = "1.2.3";
+
+            var packageName1 = "TestPackage1";
+            var packageVersion1 = "1.0.0";
+
+            using (var pathContext = new SimpleTestPathContext())
+            using (var testCertificate = new X509Certificate2(_trustedTestCert.Source.Cert))
+            {
+                // Arrange
+                EnsureVisualStudioHost();
+                var solutionService = VisualStudio.Get<SolutionService>();
+
+                solutionService.CreateEmptySolution("TestSolution", pathContext.SolutionRoot);
+                var project = solutionService.AddProject(ProjectLanguage.CSharp, projectTemplate, ProjectTargetFramework.V46, "TestProject");
+                project.Build();
+
+                Utils.CreatePackageInSource(pathContext.PackageSource, packageName1, packageVersion1);
+
+                Task.Run(async () =>
+                {
+                    var packageToSign = Utils.CreatePackage(packageName2, packageVersion2);
+                    await SignedArchiveTestUtility.CreateSignedPackageAsync(testCertificate, packageToSign, pathContext.PackageSource, packageToSign.PackageName);
+                }).GetAwaiter().GetResult();
+
+                var nugetTestService = GetNuGetTestService();
+                Assert.True(nugetTestService.EnsurePackageManagerConsoleIsOpen());
+
+                var nugetConsole = nugetTestService.GetPackageManagerConsole(project.Name);
+
+                Assert.True(nugetConsole.InstallPackageFromPMC(packageName1, packageVersion1));
+                Assert.True(nugetConsole.InstallPackageFromPMC(packageName2, packageVersion2));
+                project.Build();
+                Assert.True(VisualStudio.HasNoErrorsInErrorList());
+                Assert.True(VisualStudio.HasNoErrorsInOutputWindows());
+                Assert.True(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName1, packageVersion1));
+                Assert.True(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName2, packageVersion2));
+
+                Assert.True(nugetConsole.UninstallPackageFromPMC(packageName1));
+                Assert.True(nugetConsole.UninstallPackageFromPMC(packageName2));
+                project.Build();
+                solutionService.SaveAll();
+
+                Assert.False(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName1, packageVersion1));
+                Assert.False(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName2, packageVersion2));
+
+                nugetConsole.Clear();
+                solutionService.Save();
+            }
+        }
+
+        [CIOnlyNuGetWpfTheory]
+        [InlineData(ProjectTemplate.ClassLibrary)]
+        [InlineData(ProjectTemplate.NetCoreConsoleApp)]
+        [InlineData(ProjectTemplate.NetStandardClassLib)]
+        public void DowngradeSignedPackageToUnsignedVersionFromPMC(ProjectTemplate projectTemplate)
```

Comment:
This again feels over-kill. IMO we should try and cover all different scenarios/ edge cases via unit tests or func tests. But have E2E tests for the mainline scenarios just to make sure it works all the way end to end. We shouldn't need to cover everything through E2E tests. It will make it really slow to run all E2E tests over time.

Reply:
You are right. The test cases I will test will be:
- Install signed package
- Uninstall signed package
- Update from unsiged to signed
- Downgrade from signed to unsigned
- Install tampered signed
- Install untrusted (e.g. expired cert) signed